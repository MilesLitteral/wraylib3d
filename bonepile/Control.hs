{-# LANGUAGE AllowAmbiguousTypes #-}

module Manifest.Control where

import Manifest.Control.NetworkManager
import Manifest.Control.UDP
