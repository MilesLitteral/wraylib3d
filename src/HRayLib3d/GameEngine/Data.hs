module HRayLib3d.GameEngine.Data (
        -- module HRayLib3d.GameEngine.Data.BSP,
        -- module HRayLib3d.GameEngine.Data.GameCharacter,
        -- module HRayLib3d.GameEngine.Data.GLB,
        -- module HRayLib3d.GameEngine.Data.GLTF,
        -- module HRayLib3d.GameEngine.Data.Material,
        -- module HRayLib3d.GameEngine.Data.MD3,
        -- module HRayLib3d.GameEngine.Data.OBJ
    ) where

    -- import HRayLib3d.GameEngine.Data.BSP
    -- import HRayLib3d.GameEngine.Data.GameCharacter
    -- import HRayLib3d.GameEngine.Data.GLB
    -- import HRayLib3d.GameEngine.Data.GLTF
    -- import HRayLib3d.GameEngine.Data.Material
    -- import HRayLib3d.GameEngine.Data.MD3
    -- import HRayLib3d.GameEngine.Data.OBJ