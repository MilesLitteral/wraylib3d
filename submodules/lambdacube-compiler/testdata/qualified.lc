{-# LANGUAGE NoImplicitPrelude #-}

import qualified Prelude as P
import qualified Prelude hiding (a,b) as PH
import qualified Prelude (c,e)
-- import qualified Prelude (c,e) hiding (a,b) as PH -- fails
