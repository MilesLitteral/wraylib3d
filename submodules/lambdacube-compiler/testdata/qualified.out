main is not found
------------ desugared source code
------------ core code