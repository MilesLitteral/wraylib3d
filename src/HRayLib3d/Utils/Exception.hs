module HRayLib3d.Utils.Exception (try', catchAny) where

    import Control.Exception

    -- Misc. Exception catching code:

    --try' can be used to catch an exception in any IO function
    try' :: IO a ->  IO (Either IOException a)
    try' =  try 

    -- Avoid the "Ambiguous type variable..." error
    catchAny :: IO a -> (SomeException -> IO a) -> IO a
    catchAny = catch