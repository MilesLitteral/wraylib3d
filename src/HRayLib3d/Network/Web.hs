{-# LANGUAGE FlexibleInstances #-}
module HRayLib3d.Network.Web where 

import GHC.Exts
import Data.Maybe
import Data.Monoid
import Control.Lens 
import Data.Hashable (Hashable(..))

data (IsList a) => WebRequest a = WebRequest { request :: String, response :: a } deriving (Ord, Eq, Show)

instance Hashable (WebRequest String) where
    hashWithSalt salt lb =
        hashWithSalt salt (lb) 
    hash lb = 
        hash (lb)