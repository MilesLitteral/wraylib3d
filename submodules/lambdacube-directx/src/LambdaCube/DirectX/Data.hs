module LambdaCube.DirectX.Data where

import Control.Applicative
import Control.Monad
import Data.IORef
import Data.List as L
import Data.Maybe
import Foreign 
--import qualified Data.IntMap as IM
import qualified Data.Map as Map
import qualified Data.Set as Set
import qualified Data.Vector as V
import qualified Data.Vector.Storable as SV

--import Control.DeepSeq

import Data.Word
import Codec.Picture
import Codec.Picture.Types

import Graphics.D3D11.Interface --import Graphics.GL.Core33 
import LambdaCube.DirectX.Type
import LambdaCube.DirectX.Util

-- Buffer
disposeBuffer :: ID3D11Buffer -> IO ()
disposeBuffer (Buffer _ bo) = withArray [bo] $ glDeleteBuffers 1

compileBuffer :: [SimpleVertex] -> IO ID3D11Buffer
compileBuffer arrs = do
    let bd = D3D11BufferDesc
        { byteWidth = fromIntegral $ 8 * sizeOf (undefined :: SimpleVertex)
        , usage = D3D11UsageDefault
        , bindFlags = d3d11BindFlags [D3D11BindVertexBuffer]
        , cpuAccessFlags = 0
        , miscFlags = 0
        , structureByteStride = 0 }
    buffer <- createBuffer device bd arrs                  
    iaSetVertexBuffers deviceContext 0 [(buffer, fromIntegral $ sizeOf (undefined :: SimpleVertex), 0)]
    return $! ID3D11Buffer buffer

-- updateBuffer :: ID3D11Buffer -> [(Int,Array)] -> IO ()
-- updateBuffer (Buffer arrDescs bo) arrs = do
--     glBindBuffer GL_ARRAY_BUFFER bo
--     forM arrs $ \(i,Array arrType cnt setter) -> do
--         let ArrayDesc ty len offset size = arrDescs V.! i
--         when (ty == arrType && cnt == len) $
--             setter $! glBufferSubData GL_ARRAY_BUFFER (fromIntegral offset) (fromIntegral size)
--     glBindBuffer GL_ARRAY_BUFFER 0

-- bufferSize :: ID3D11Buffer -> Int
-- bufferSize = V.length . bufArrays

-- arraySize :: ID3D11Buffer -> Int -> Int
-- arraySize buf arrIdx = arrLength $! bufArrays buf V.! arrIdx

-- arrayType :: ID3D11Buffer -> Int -> ArrayType
-- arrayType buf arrIdx = arrType $! bufArrays buf V.! arrIdx

-- Texture
disposeTexture :: TextureData -> IO ()
disposeTexture (TextureData to) = withArray [to] $ glDeleteTextures 1

-- FIXME: Temporary implemenation
uploadTexture2DToGPU :: DynamicImage -> IO TextureData
uploadTexture2DToGPU = uploadTexture2DToGPU' True False True False

uploadTexture2DToGPU' :: ID3D11Device -> DynamicImage -> IO TextureData
uploadTexture2DToGPU' device bitmap' = do
    let bitmap = case bitmap' of
          ImageRGB8   i@(Image w h _) -> bitmap'
          ImageRGBA8  i@(Image w h _) -> bitmap'
          ImageYCbCr8 i@(Image w h _) -> ImageRGB8 $ convertImage i
          di -> ImageRGBA8 $ convertRGBA8 di
        (width,height) = bitmapSize bitmap
            bitmapSize (ImageRGB8  (Image w h _)) = (w,h)
            bitmapSize (ImageRGBA8 (Image w h _)) = (w,h)
            bitmapSize _ = error "unsupported image type :("
            withBitmap (ImageRGB8  (Image w h v)) f = SV.unsafeWith v $ f (w,h) 3 0
            withBitmap (ImageRGBA8 (Image w h v)) f = SV.unsafeWith v $ f (w,h) 4 0
            withBitmap _ _ = error "unsupported image type :("
        td = D3D11Texture2DDesc
                width
                height
                1
                1
                DxgiFormatD24UnormS8Uint
                (DxgiSampleDesc 1 0)
                D3D11UsageDefault
                (d3d11BindFlags []) --D3D11BindDepthStencil
                0
                0
    to <- createTexture2D device td (SV.toList v :: [Word16])
    return $ TextureData to
