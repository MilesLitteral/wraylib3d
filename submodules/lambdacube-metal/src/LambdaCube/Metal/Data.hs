module LambdaCube.Metal.Data where

import Control.Applicative
import Control.Monad
import Data.IORef
import Data.List as L
import Data.Maybe
import Foreign 
--import qualified Data.IntMap as IM
import qualified Data.Map as Map
import qualified Data.Set as Set
import qualified Data.Vector as V
import qualified Data.Vector.Storable as SV

--import Control.DeepSeq

import Data.Word
import Codec.Picture
import Codec.Picture.Types

import LambdaCube.Metal.Type
import LambdaCube.Metal.Util
import Graphics.GL.Core33
--import LambdaCube.Metal.Bindings  --Graphics.GL.Core33
--Utility Functions
calcDesc (offset,setters,descs) (Array arrType cnt setter) =
          let size = cnt * sizeOfArrayType arrType
            in (size + offset, (offset, size,setter):setters, ArrayDesc arrType cnt offset size:descs)

-- Buffer
disposeBuffer :: Buffer -> IO ()
disposeBuffer (Buffer _ bo) = withArray [bo] $ resetCommandsInBuffer 1 

compileBuffer :: MTLDevice -> [Array] -> IO Buffer
compileBuffer dev arrs = do
    let (bufSize,arrSetters,arrDescs) = foldl' calcDesc (0,[],[]) arrs
    bo <- alloca $! \pbo -> makeBuffer dev pbo 1 MTLResourceOptionCPUCacheModeDefault >> peek pbo --glGenBuffers
    forM_ arrSetters $! \(offset,size,setter) -> setter $! makeBuffer dev (fromIntegral offset) (fromIntegral size) MTLResourceOptionCPUCacheModeDefault
    return $! Buffer (V.fromList $! reverse arrDescs) bo

updateBuffer :: MTLDevice -> Buffer -> [(Int,Array)] -> IO ()
updateBuffer (Buffer arrDescs bo) arrs = forM arrs $ 
    \(i,Array arrType cnt setter) -> do
        let ArrayDesc ty len offset size = arrDescs V.! i
        when (ty == arrType && cnt == len) $ 
            setter $! makeBuffer dev (fromIntegral offset) (fromIntegral size) MTLResourceOptionCPUCacheModeDefault

bufferSize :: Buffer -> Int
bufferSize = V.length . bufArrays

arraySize :: Buffer -> Int -> Int
arraySize buf arrIdx = arrLength $! bufArrays buf V.! arrIdx

arrayType :: Buffer -> Int -> ArrayType
arrayType buf arrIdx = arrType $! bufArrays buf V.! arrIdx

-- Texture
disposeTexture :: TextureData -> IO ()
disposeTexture (TextureData to) = withArray [to] $ metalDeleteTextures --glDeleteTextures 1

-- FIXME: Temporary implemenation
uploadTexture2DToGPU :: DynamicImage -> IO TextureData
uploadTexture2DToGPU = uploadTexture2DToGPU' True False True False

uploadTexture2DToGPU' :: Bool -> DynamicImage -> IO TextureData
uploadTexture2DToGPU' isMip bitmap' = do
    let bitmap = case bitmap' of
                    ImageRGB8   i@(Image w h _) -> bitmap'
                    ImageRGBA8  i@(Image w h _) -> bitmap'
                    ImageYCbCr8 i@(Image w h _) -> ImageRGB8 $ convertImage i
                    di -> ImageRGBA8 $ convertRGBA8 di
        (width,height) = bitmapSize bitmap
                         bitmapSize (ImageRGB8  (Image w h _)) = (w,h)
                         bitmapSize (ImageRGBA8 (Image w h _)) = (w,h)
                         bitmapSize _ = error "unsupported image type :("
                         withBitmap (ImageRGB8 (Image w h v))  f = SV.unsafeWith v $ f (w,h) 3 0
                         withBitmap (ImageRGBA8 (Image w h v)) f = SV.unsafeWith v $ f (w,h) 4 0
                         withBitmap _ _ = error "unsupported image type :("
        to <- alloca $! \pto -> newTextureWithDescriptor pto >> MTLTextureDescriptor PixelFormatBGRA8Unorm width height --peek pto
        replaceRegion 0 v $ castPtr ptr
    when isMip $ replaceRegion 1 v $ castPtr ptr
    return $ TextureData to
