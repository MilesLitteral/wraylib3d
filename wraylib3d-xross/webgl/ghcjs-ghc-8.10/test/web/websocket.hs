{-# LANGUAGE OverloadedStrings, ScopedTypeVariables #-}
module Main where

import           Control.Concurrent
import           Control.Exception

import qualified Data.JSString as JSS
import           Data.Monoid

import           JavaScript.Web.Location
import           JavaScript.Web.WebSocket (WebSocketRequest(..), WebSocket)
import qualified JavaScript.Web.WebSocket as WS
import           JavaScript.Web.MessageEvent (MessageEvent)
import qualified JavaScript.Web.MessageEvent as ME
import           JavaScript.Web.CloseEvent (CloseEvent)
import qualified JavaScript.Web.CloseEvent as CE

import           JavaScript.TypedArray.ArrayBuffer (ArrayBuffer)

main :: IO ()
main = test1 >> test2 >> test3

-- text messages, close from client
test1 :: IO ()
test1 = do
  putStrLn "test1"
  ws <- connectWS True
  WS.send "-5" ws
  threadDelay 1000000
  WS.send "5" ws
  threadDelay 1000000
  WS.close Nothing Nothing ws
  threadDelay 1000000

test2 :: IO ()
test2 = do
  putStrLn "test2"
  ws  <- connectWS True
  buf <- js_makeArrayBuffer
  WS.sendArrayBuffer buf ws
  WS.send "0" ws
  threadDelay 1000000

test3 :: IO ()
test3 = do
  putStrLn "test3"
  (connectWS False >>= WS.close Nothing Nothing) `catch` \(e::SomeException) -> print e
  threadDelay 1000000

connectWS :: Bool -> IO WebSocket
connectWS valid = do
  host <- getHost =<< getWindowLocation
  WS.connect $ WebSocketRequest
    { url       = "ws://" <> (if valid then host else "nonexistenthost:62312") <> "/ws"
    , protocols = []
    , onClose   = Just showClose
    , onMessage = Just showMessage
    }

showMessage :: MessageEvent -> IO ()
showMessage msg = putStrLn ("received message: " ++ str)
  where
    str = case ME.getData msg of
      ME.StringData s      -> "string: " ++ JSS.unpack s
      ME.BlobData _        -> "Blob"
      ME.ArrayBufferData _ -> "ArrayBuffer"

showClose :: CloseEvent -> IO ()
showClose ce = putStrLn $
  "connection closed, status: " ++ show (CE.getCode ce) ++
  " clean: " ++ show (CE.wasClean ce) ++
  " reason: " ++ show (JSS.unpack $ CE.getReason ce)

{-
printWebSocketInfo :: WebSocket -> IO ()
printWebSocketInfo ws = do
  let pi n xs = putStrLn (n ++ " " ++ JSS.unpack xs)
  pi "protocol"   =<< WS.getProtocol   ws
  pi "extensions" =<< WS.getExtensions ws
  putStrLn . ("buffered amount "++) . show =<< WS.getBufferedAmount ws
  pi "url" (WS.getUrl ws)
  putStrLn . ("readystate "++) . show =<< WS.getReadyState ws
  putStrLn . ("binary type "++) . show =<< WS.getBinaryType ws
-}
-- fixme use JavaScript.TypedArray API when it's done
foreign import javascript unsafe
  "var x = new Int8Array(2); x[0] = 1; x[1] = 2; $r = x.buffer;"
  js_makeArrayBuffer :: IO ArrayBuffer
