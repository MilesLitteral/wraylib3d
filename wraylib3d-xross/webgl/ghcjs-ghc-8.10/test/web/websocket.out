test1
received message: string: TestTextMessage
received message: string: TestTextMessage
received message: string: TestTextMessage
received message: string: TestTextMessage
received message: string: TestTextMessage
received message: ArrayBuffer
received message: ArrayBuffer
received message: ArrayBuffer
received message: ArrayBuffer
received message: ArrayBuffer
connection closed, status: 1000 clean: True reason: ""
test2
connection closed, status: 1000 clean: True reason: ""
test3
user error (WebSocket failed to connect)
